plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/storage/TrieBasedStore.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.wlm.rule.storage;
+
+import org.apache.commons.collections4.trie.PatriciaTrie;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This is a patricia trie based implementation of AttributeValueStore
+ */
+public class TrieBasedStore implements AttributeValueStore {
+    PatriciaTrie<String> trie;
```

Comment:
Does `PatriciaTrie` allow <K,V>? I am thinking we can have <String, QueryGroup> instead of <String, String> to not having to resolve the `QueryGroup` again, since it will be available while adding to Trie

Reply:
Maybe, we can even assume `String` as the default key type since we are dealing with `Attributes`, but should be generic for the value

Reply:
Sure, PatriciaTrie specifically does not support generic key types but only value types.